### PR TITLE
Open file to edit from debugger

### PR DESCRIPTION
**Pull request by gaborvecsei**
Issue: #211 

## Feature

When you are on a file at the "stack" window, you can press `Ctrl+e` and the file will open with your editor, based on the `$EDITOR` environment variable. When you are done with editing you can just save and quit the file, and you will be returned to the debugger.

`reload()` is not used (because of this comment: https://github.com/inducer/pudb/issues/211#issuecomment-327031721), so the edits won't be noticable immediately, only after quitting the debugger and starting it again

### CODE REVIEW

File: pudb/debugger.py [hunk 1]
```diff
@@ -342,6 +342,22 @@ def set_frame_index(self, index):
...
 
+    def open_file_to_edit(self, index):
+        self.curframe, line_number = self.stack[index]
+        filename = self.curframe.f_code.co_filename
+
+        editor = os.environ.get("EDITOR", "nano")
+        editor = editor.lower()
+
+        editor_command = f"{editor}"
+        if editor in {"vi", "vim", "nvim"}:
```
Comment (asmeurer) [inline pudb/debugger.py:353]: I believe emacs supports this as well.

**Comment by asmeurer** [comment]
This needs to be added to the help. Maybe also with a note that changes will only take effect when pudb is restarted.

File: pudb/debugger.py (hunk 1 above)
Comment (gaborvecsei) [reply pudb/debugger.py:353]: Good catch I'll add it!

Comment (gaborvecsei) [reply pudb/debugger.py]: Done

**Comment by gaborvecsei** [comment]
> This needs to be added to the help. Maybe also with a note that changes will only take effect when pudb is restarted.

@asmeurer I'll add it. The question is where should I put this? Under `Features` in the README?

**Comment by asmeurer** [comment]
You can add it to the docs if you want, but I actually meant the interactive help that currently lists every keyboard command (what pops when you press `?`).

File: pudb/debugger.py [hunk 2]
```diff
@@ -342,6 +342,22 @@ def set_frame_index(self, index):
...
+        filename = self.curframe.f_code.co_filename
+
+        editor = os.environ.get("EDITOR", "nano")
+        editor = editor.lower()
+
+        editor_command = f"{editor}"
+        if editor in {"vi", "vim", "nvim", "emacs"}:
+            editor_command = f"{editor} +{line_number}"
+        elif editor == "nano":
+            editor_command = f"{editor} -l {line_number}"
```
Comment (asmeurer) [inline pudb/debugger.py:356]: According to `man nano` at least on my machine, nano also supports the `+line_number` syntax. Maybe we should just always do this regardless of the editor. I suspect other editors like vscode will probably also support this.

Comment (inducer) [inline pudb/debugger.py:356]: This fails if an explicit path is specified, or if the command is `emacs-gtk` or some-such. Maybe check for substrings? It might also be useful to let the user customize the invocation in settings.

File: pudb/debugger.py [hunk 3]
```diff
@@ -342,6 +342,22 @@ def set_frame_index(self, index):
...
         self.ui.stack_list._w.set_focus(self.ui.translate_ui_stack_index(index))
 
+    def open_file_to_edit(self, index):
+        self.curframe, line_number = self.stack[index]
+        filename = self.curframe.f_code.co_filename
+
+        editor = os.environ.get("EDITOR", "nano")
+        editor = editor.lower()
+
+        editor_command = f"{editor}"
```
Comment (inducer) [inline pudb/debugger.py:352]: Could you move this to the `else` branch?

File: pudb/debugger.py [hunk 4]
```diff
@@ -342,6 +342,22 @@ def set_frame_index(self, index):
 
         self.ui.stack_list._w.set_focus(self.ui.translate_ui_stack_index(index))
 
+    def open_file_to_edit(self, index):
+        self.curframe, line_number = self.stack[index]
+        filename = self.curframe.f_code.co_filename
```
Comment (inducer) [inline pudb/debugger.py:347]: AFAIK, this may be None. Show an appropriate message if that's the case, instead of crashing.

File: pudb/debugger.py [hunk 5]
```diff
@@ -342,6 +342,22 @@ def set_frame_index(self, index):
...
+        editor = editor.lower()
+
+        editor_command = f"{editor}"
+        if editor in {"vi", "vim", "nvim", "emacs"}:
+            editor_command = f"{editor} +{line_number}"
+        elif editor == "nano":
+            editor_command = f"{editor} -l {line_number}"
+
+        command_to_run = f"{editor_command} {filename}"
+        os.system(command=command_to_run)
```
Comment (inducer) [inline pudb/debugger.py:359]: This will fail if the filename contains spaces. Use `subprocess` and don't invoke via the shell.

File: pudb/debugger.py [hunk 6]
```diff
@@ -342,6 +342,22 @@ def set_frame_index(self, index):
 
         self.ui.stack_list._w.set_focus(self.ui.translate_ui_stack_index(index))
 
+    def open_file_to_edit(self, index):
+        self.curframe, line_number = self.stack[index]
+        filename = self.curframe.f_code.co_filename
+
+        editor = os.environ.get("EDITOR", "nano")
+        editor = editor.lower()
```
Comment (inducer) [inline pudb/debugger.py:350]: I'd remove this. Unix file systems can be case-sensitive, and somebody might have an upper-case letter in their `$EDITOR`.

File: pudb/debugger.py (hunk 4 above)
Comment (inducer) [inline pudb/debugger.py:347]: It may also not be a valid filename. Maybe check before proceeding?

File: pudb/debugger.py (hunk 5 above)
Comment (inducer) [inline pudb/debugger.py:359]: What completion semantics do we expect here? Should the source be reloaded when we come back from the editor? (This involves two possible parts: `linecache` and Python module reload) (If we do nothing, it'll be stale, which will be confusing.)

**Review by inducer — COMMENTED** [review]
Thanks for working on this. A few thoughts below.

**Comment by inducer** [comment]
To cut down on noise in my email, I'm unsubscribing from this PR. When it next needs my attention, please @-mention me or hit the "request review" button. Otherwise, I may not see your messages in a timely manner.

File: pudb/debugger.py (hunk 5 above)
Comment (asmeurer) [reply pudb/debugger.py:359]: Yeah, I'm unsure about this too. Doing a `reload` is probably a bad idea, as it is wrong even more subtle ways (if you really want to do that, you can do it yourself manually in the console). Maybe after editing it should popup a message asking if you want to stop the debugger so you can restart the interpreter?

File: pudb/debugger.py (hunk 2 above)
Comment (gaborvecsei) [reply pudb/debugger.py]: DONE

File: pudb/debugger.py (hunk 3 above)
Comment (gaborvecsei) [reply pudb/debugger.py]: Because of a request above this is not needed.

Every editor supports the `+l flag` so we can just `f"{editor} +l {line_number} {filename}"`

File: pudb/debugger.py (hunk 4 above)
Comment (gaborvecsei) [reply pudb/debugger.py:347]: Done

File: pudb/debugger.py (hunk 5 above)
Comment (gaborvecsei) [reply pudb/debugger.py]: I think not using `reload` is the better option. I will display a message, so the user is aware that the file changed, and they can stop the execution and restart if they want

Comment (gaborvecsei) [reply pudb/debugger.py]: DONE

**Comment by gaborvecsei** [comment]
Feature is added to the help (when pressing `?`)

**Comment by gaborvecsei** [comment]
Hi @inducer, @asmeurer!
Thank you for the look at the PR.
I am done with all the requests, can you review the PR?

File: pudb/debugger.py [hunk 7]
```diff
@@ -123,6 +123,7 @@
 
 Keys in stack list:
     enter - jump to frame
+    Ctrl-e - open file at line to edit with $EDITOR
```
Comment (asmeurer) [inline pudb/debugger.py:126]: Oh I didn't notice that this is based on the stack list. I was expecting this to be implemented in the main code view.

Comment (gaborvecsei) [reply pudb/debugger.py:126]: I understand why you think it would make more sense. I also thought about this and I implemented it here because of the resons:
- Maybe the developer would like to "look around" in the stack list for other files which are "effected" in the debugging. And it is easier with directly opening the files from stack list
- The `Ctrl-e` shortcut is useable/not allocated
- No (new) users can accidentaly open the file editor window. It takes 1 more shortcut to go to the stack list
- It is not confusing (even with the warning message) that the file is not changed which is in front of them after the edit.
- If we would open the current file which line would we jump to? The one under our cursor, the top/bottom/middle breakpoint? --> this can make some consusion, while on stack list the line number is displayed so no surprises will happen for the user

(maybe I am biased by my experience 😄 )

Comment (inducer) [reply pudb/debugger.py:126]: FWIW, I also didn't catch that this was on the stack list. So there's at least two data points on where folks would expect this functionality to live. (I'm not saying remove it from the stack, but also add it to the main source window, too.) As for what line to show, I think the current highlight is the obvious choice.

File: pudb/debugger.py (hunk 5 above)
Comment (inducer) [reply pudb/debugger.py]: You could *offer* a reload in the message. Do you think that's reasonable?

File: pudb/debugger.py (hunk 7 above)
Comment (gaborvecsei) [reply pudb/debugger.py:127]: I fixed it, now both on Main view and Stack view the `Ctrl-e` edit functionality is working
@inducer Can you review it?

File: pudb/debugger.py [hunk 8]
```diff
@@ -1103,6 +1123,35 @@ def examine_frame(w, size, key):
...
+                self.message("Exception happened when trying to edit the file:"
+                             "\n\n%s" % ("".join(format_exception(sys.exc_info()))),
+                    title="File Edit Error")
+                return
+
+            self.message("File is changed, but the execution is continued with"
+                         " the 'old' codebase.\n"
+                         f"Changed file: {filename_edited}\n\n"
+                         "Please quit and restart to see changes",
+                         title="File is changed")
```
Comment (inducer) [inline pudb/debugger.py:1141]: This can claim that the code was changed, when it didn't, which might spook users. Maybe check the mtime with `stat`?

File: pudb/debugger.py [hunk 9]
```diff
@@ -342,6 +344,24 @@ def set_frame_index(self, index):
 
         self.ui.stack_list._w.set_focus(self.ui.translate_ui_stack_index(index))
 
+    def open_file_to_edit(self, filename, line_number):
```
Comment (inducer) [inline pudb/debugger.py:347]: No real reason for this to be a method. It doesn't use `self` AFAICT. Move to a function?

File: pudb/debugger.py [hunk 10]
```diff
@@ -342,6 +344,24 @@ def set_frame_index(self, index):
...
+
+        if not filename:
+            raise ValueError("Could not get filename as it was None")
+
+        if not os.path.isfile(filename):
+            raise FileNotFoundError(f"Could not find the file with name: {filename}")
```
Comment (inducer) [inline pudb/debugger.py:353]:
```suggestion
            raise FileNotFoundError(f"'{filename}' not found or is not a file.")
```

File: pudb/debugger.py [hunk 11]
```diff
@@ -342,6 +344,24 @@ def set_frame_index(self, index):
 
         self.ui.stack_list._w.set_focus(self.ui.translate_ui_stack_index(index))
 
+    def open_file_to_edit(self, filename, line_number):
+
+        if not filename:
+            raise ValueError("Could not get filename as it was None")
```
Comment (inducer) [inline pudb/debugger.py:350]: Make this more actionable for a user, explaining what may have happened.

File: pudb/debugger.py [hunk 12]
```diff
@@ -1103,6 +1123,35 @@ def examine_frame(w, size, key):
...
+
+        def open_editor_on_stack_frame(w, size, key):
+            _, pos = self.stack_list._w.get_focus()
+            index = self.translate_ui_stack_index(pos)
+
+            curframe, line_number = self.debugger.stack[index]
+            file_name = curframe.f_code.co_filename
+
+            open_file_editor(file_name, line_number)
+            redraw_screen(w, size, key)
```
Comment (inducer) [inline pudb/debugger.py:1151]: Why is this necessary? (Instead, you should stop/start the Urwid screen, which may switch the terminal between front and alternate buffers.)

**Review by inducer — COMMENTED** [review]
Thanks for following up. Some more thoughts below.

File: pudb/debugger.py (hunk 8 above)
Comment (gaborvecsei) [reply pudb/debugger.py:1141]: And what would be the threshold where we can be certain there was a change? 1minute,  5 minutes?
Even if it is 30 minutes, sometimes I modify code for >30mins, even drink coffee, etc.

I just saying we cannot expect the users to quickly modify code when they are debugging.

What if the message would say something like

> You were in edit mode. If anything was changed in the code, please restart the debugging session, if not, you can continue in this session.

File: pudb/debugger.py (hunk 9 above)
Comment (gaborvecsei) [reply pudb/debugger.py:347]: DONE

File: pudb/debugger.py (hunk 11 above)
Comment (gaborvecsei) [reply pudb/debugger.py:350]: Actually as I see this can not happen.
I removed the raising of the exception

Comment (gaborvecsei) [reply pudb/debugger.py:350]: DONE

File: pudb/debugger.py (hunk 12 above)
Comment (gaborvecsei) [reply pudb/debugger.py]: We don't need it 🙂  Good catch!

Comment (gaborvecsei) [reply pudb/debugger.py]: DONE

**Comment by gaborvecsei** [comment]
@inducer I am done with the requested fixes, and for the "changed file popup" I left an answer.

File: pudb/debugger.py (hunk 8 above)
Comment (inducer) [reply pudb/debugger.py:1138]: I'm not sure I understand. If the mtime is different than before we started the editor, it's pretty clear that the file was changed.

File: pudb/debugger.py [hunk 13]
```diff
@@ -2000,6 +2045,12 @@ def redraw_screen(w, size, key):
         def help(pages):
             self.message(pages, title="PuDB - The Python Urwid Debugger")
 
+        def edit_current_frame(w, size, key):
+            file_name = self.debugger.curframe.f_code.co_filename
+            line_number = self.debugger.curframe.f_lineno
+            open_file_editor(file_name, line_number)
+            redraw_screen(w, size, key)
```
Comment (inducer) [inline pudb/debugger.py:2052]: Elsewhere, I thought we got rid of `redraw_screen`?

Comment (gaborvecsei) [reply pudb/debugger.py:2052]: Sorry I missed this, and removed only 1 from another place.
DONE

File: pudb/debugger.py (hunk 8 above)
Comment (gaborvecsei) [reply pudb/debugger.py]: Sorry, I overthinked this one. Now it is implemented as requested.

DONE

File: pudb/debugger.py (hunk 5 above)
Comment (gaborvecsei) [reply pudb/debugger.py]: (This conversation is outdated)

**Comment by gaborvecsei** [comment]
@inducer Requests are fixed, please review

**Comment by inducer** [comment]
This looks good to me. As you can see, I changed the behavior for Ctrl-e in the source window to use the focused rather than the current line. I hope you're OK with that. Thanks for your contribution!

**Comment by gaborvecsei** [comment]
> This looks good to me. As you can see, I changed the behavior for Ctrl-e in the source window to use the focused rather than the current line. I hope you're OK with that. Thanks for your contribution!

Of course, I am okay with that. Thank you very much for the reviews!
